### PR TITLE
TypeVisitor to emit diagnostics

### PR DESCRIPTION
**Pull request by srividya-sundaram**

### CODE REVIEW

File: clang/test/SemaSYCL/kernelname-enum.cpp [hunk 1]
```diff
@@ -54,24 +89,12 @@ int main() {
...
-    cgh.single_task(f2);
-  });
-
-  q.submit([&](cl::sycl::handler &cgh) {
-    cgh.single_task(f3);
-  });
-
-  q.submit([&](cl::sycl::handler &cgh) {
-    cgh.single_task(f4);
-  });
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/kernelname-enum.cpp]: Why did you remove them?

File: clang/test/SemaSYCL/kernelname-enum.cpp [hunk 2]
```diff
@@ -1,13 +1,42 @@
...
+
+#endif
 
 enum unscoped_enum_int : int {
   val_1,
   val_2
 };
 
-// expected-note@+1 {{'unscoped_enum_no_type_set' declared here}}
+// expected-note@+1 1+ {{'unscoped_enum_no_type_set' declared here}}
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/kernelname-enum.cpp:11]: What is "1+"? I've never seen this before.

File: clang/test/SemaSYCL/kernelname-enum.cpp [hunk 3]
```diff
@@ -1,13 +1,42 @@
-// RUN: %clang_cc1 -I %S/Inputs -fsycl -fsycl-is-device -fsycl-int-header=%t.h -fsyntax-only -Wno-sycl-2017-compat -verify %s
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -fsycl-int-header=%t.h -fsyntax-only -Wno-sycl-2017-compat -verify %s
 
-#include "sycl.hpp"
+//#include "Inputs/sycl.hpp"
+#ifndef SYCL_HPP
+#define SYCL_HPP
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/kernelname-enum.cpp]: Don't think that you need it.

File: clang/test/SemaSYCL/kernelname-enum.cpp [hunk 4]
```diff
@@ -1,13 +1,42 @@
...
+}
+  template <typename KernelName = auto_name, typename KernelType>
+  void single_task(const KernelType &kernelFunc) {
+    using NameT = typename get_kernel_name_t<KernelName, KernelType>::name;
+#ifdef __SYCL_DEVICE_ONLY__
+// expected-error@+1 1+{{kernel name is invalid. Unscoped enum requires fixed underlying type}}
+    kernel_single_task<NameT>(kernelFunc);
+#else
+    kernelFunc();
+#endif
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/kernelname-enum.cpp]: We are testing for device, I think you don't need it.
```suggestion
```

**Review by Fznamznon — COMMENTED** [review]
Just ignore my comments about single_task function if catching diagnostics from another file works.

File: clang/test/SemaSYCL/kernelname-enum.cpp (hunk 2 above)
Comment (srividya-sundaram) [reply clang/test/SemaSYCL/kernelname-enum.cpp:11]: It is used to indicate the number of times this error is expected. In our case it is 2, since I invoke the kernel-caller-function twice for f2 and f5

Comment (srividya-sundaram) [reply clang/test/SemaSYCL/kernelname-enum.cpp:11]: From : https://clang.llvm.org/doxygen/classclang_1_1VerifyDiagnosticConsumer.html
Where the diagnostic is expected to occur a minimum number of times, this can be specified by appending a '+' to the number. Example:

void f(); // expected-note 0+ {{previous declaration is here}}
void g(); // expected-note 1+ {{previous declaration is here}}

Comment (Fznamznon) [reply clang/test/SemaSYCL/kernelname-enum.cpp:11]: Could you please specify it explicitly? i.e. just '2'
